### PR TITLE
Combine Duplicate IPv4/IPv6 TCP code

### PR DESCRIPTION
**Pull request by HTRamsey**
Remove duplicate TCP code

Description
-----------
There's a few areas of duplicate code, and this was  the largest with around 300 duplicate lines. Use a helper that makes this function agnostic to the frame type. This makes testing, verification, and maintenance of code simpler over time.

Test Steps
-----------

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by paulbartell** [comment]
Thanks for the PR. I'll send this over to a FreeRTOS+TCP expert for review.

**Comment by HTRamsey** [comment]
This can be simplified (especially FreeRTOS_TCP_Transmission.c) after ARP/ND updates.

**Comment by HTRamsey** [comment]
@tony-josi-aws I have simplified this as well to only address the place with 300 identical lines. Sorry for the larger PRs, I will make them smaller and simpler.

**Review by tony-josi-aws — APPROVED** [review]

File: source/FreeRTOS_TCP_IP.c [hunk 1]
```diff
@@ -648,36 +691,244 @@
...
+                /* pxSocket is not NULL when xResult != pdFAIL. */
+                configASSERT( pxSocket != NULL ); /* LCOV_EXCL_LINE ,this branch will not be hit*/
+
+                /* Touch the alive timers because we received a message for this
+                 * socket. */
+                prvTCPTouchSocket( pxSocket );
+
+                /* Parse the TCP option(s), if present. */
+
+                /* _HT_ : if we're in the SYN phase, and peer does not send a MSS option,
```
Comment (moninom1) [inline source/FreeRTOS_TCP_IP.c:874]: Can we remove _HT_ ?

**Review by moninom1 — APPROVED** [review]

**Review by shubnil — APPROVED** [review]

**Pull request merged**